File: core/java/src/main/java/hara/lang/protocol/IStreamDuplex.java
package hara.lang.protocol;

import hara.lang.declaration.HaraAvailability;
import hara.lang.declaration.HaraProtocolBinding;

/** Bidirectional stream with graceful and exceptional termination. */
@HaraProtocolBinding(
    namespace = "std.protocol.istreamduplex",
    name = "IStreamDuplex",
    parents = {"IStream", "IStreamWrite", "IAbort"},
    availability = HaraAvailability.CAPABILITY_GATED,
    capability = "native-runtime-protocols")
public interface IStreamDuplex extends IStream, IStreamWrite, IAbort {}

File: core/rust/src/lang/protocol/istreamduplex.rs
use super::{IAbort, IStream, IStreamWrite};
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.istreamduplex",
    name = "IStreamDuplex",
    parents = ["IStream", "IStreamWrite", "IAbort"],
    availability = "capability-gated",
    capability = "native-runtime-protocols"
)]
pub trait IStreamDuplex: IStream + IStreamWrite + IAbort {}
